### PR TITLE
chore: harden repo skill mirrors and setup verification

### PR DESCRIPTION
**Pull request by cobmojo**
## Summary

- README: document all skill-related scripts (`skills:sync`, `skills:verify`, `skills:refresh-upstream`) in a table, plus the maintainer workflow for vendored Supabase skills (`npx skills add` → `skills:refresh-upstream` → sync/verify). Note `setup:verify` 401 = mismatched Supabase URL/anon key.
- Prior commit on branch: Windows `setup.ps1` order aligned with Unix `setup`; README quickstart clarified.

## Verification

- `bash -n scripts/setup/index.sh`
- `node --check` on `sync-agent-skills.mjs`, `verify-skills-sync.mjs`, `refresh-upstream-skills.mjs`, `setup-verify.mjs`
- `bun run skills:verify` (pass)
- `bun run skills:refresh-upstream` (pass)
- `bun run setup:verify` fails in this sandbox with HTTP 401 (expected when `.env.local` anon key does not match project URL)

<div><a href="https://cursor.com/agents/bc-2ca6fa76-3de6-4f8f-ba18-1dad461a8771"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-2ca6fa76-3de6-4f8f-ba18-1dad461a8771"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

<h3>Greptile Summary</h3>

This PR hardens skill mirror management: it adds a manifest-based prune step to `sync-agent-skills.mjs` (so removed canonical skills are deleted from mirrors), extends setup scripts to fail-fast on mirror drift, ships the `react-doctor` skill, and documents the full skill-management workflow. It directly addresses the path-traversal concern from the prior review, adding a regex validator, a path-containment check, and a subprocess-level test.

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — all remaining findings are P2 style/diagnostic improvements that do not affect correctness or security.

The path-traversal concern from the previous review is fully addressed: the regex guard, path-containment check, and the re-wrapping in readCanonicalManifest together ensure no rm is ever called with an unsafe path. The one P2 comment is about diagnostic clarity when that guard fires, not about whether it fires. No P0/P1 issues remain.

scripts/sync-agent-skills.mjs — the readCanonicalManifest error-wrapping pattern is the only item worth a follow-up look.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| scripts/sync-agent-skills.mjs | Adds manifest-based prune + path-traversal hardening; one diagnostic concern: assertSafeCanonicalSkillDirName errors thrown inside readCanonicalManifest are caught and re-wrapped, obscuring the root cause in the top-level message. |
| tests/unit/script-verifiers.test.ts | New traversal test runs the script as a subprocess and matches on the cause-chain string in stderr; functionally correct on Bun/Node 18+ but test implicitly depends on runtime printing the [cause] chain. |
| scripts/setup/index.sh | Correctly inserts skills:verify after bun install and before setup:verify. |
| scripts/setup.ps1 | Mirrors Unix setup order: skills:verify added outside the -SkipInstall block (intentional — skill verification doesn't require installed deps) and before setup:verify. |
| docs/ai/skills/react-doctor/SKILL.md | New canonical skill file; content is complete with Triggers/Checklist sections and matches the mirrors. |
| README.md | Accurate documentation of skill scripts table, vendored Supabase skill refresh workflow, and setup:verify 401 hint. |
| .agents/skills/.repo-canonical-skills.json | New manifest file listing all 19 canonical skills; content matches docs/ai/skills/ directory inventory. |
| AGENTS.md | Adds react-doctor skill routing entry pointing at the new canonical docs path. |

</details>

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[main] --> B[listCanonicalSkillsForSync\ndocs/ai/skills/]
    B --> |assertSafeCanonicalSkillDirName| C{for each targetRoot}
    C --> D[pruneStaleCanonicalSkills\nread old manifest → diff → rm stale]
    D --> |assertSafeCanonicalSkillDirName\nassertMirrorSkillDirUnderRoot| E[syncCanonicalSkill\noverlayDirectory]
    E --> F[writeCanonicalManifest\n.repo-canonical-skills.json]
    F --> G[listAgentSkillsForMirror\n.agents/skills/]
    G --> |assertSafeCanonicalSkillDirName| H[mirrorAgentSkillToCursor\n.agents → .cursor]
    H --> I[Done]

    style D fill:#f9c,stroke:#c66
    style F fill:#cfc,stroke:#6a6
```
</details>

<sub>Reviews (3): Last reviewed commit: ["Refactor skill directory name validation..."](https://github.com/asymmetric-al/core/commit/f615142adb081eafec98d1e2e3d8ce2f5d5cde18) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28129054)</sub>

### CODE REVIEW

**Review by musab03 — APPROVED** [review]
What this PR does:
This is a pure tooling and documentation chore with no application runtime, API, or database changes. The work covers four areas.

1. Canonical skill manifest
A committed `.repo-canonical-skills.json` manifest was introduced under `.agents/skills/` and `.cursor/skills/. This` gives the sync script a source of truth for which skill directories are canonical, enabling safe pruning of mirror directories that no longer correspond to a canonical skill.

2. Mirror pruning and sync hardening
`scripts/sync-agent-skills.mjs` was extended with prune logic that removes stale canonical mirror directories. The script now validates skill names as single-segment slugs before any `rm `operation and checks that resolved paths stay under the mirror root, preventing path traversal on corrupt or hand-edited manifest entries.

3. Setup and CI wiring
`bun run skills:verify` is now wired into Unix and Windows setup scripts after install, so mirror drift is caught at onboarding time rather than discovered later.

4. Documentation and skill promotion
`react-doctor` was promoted into `docs/ai/skills/` as a canonical skill with mirrors updated. README, playbook, and `find-skills` text were updated to explain the canonical vs mirror workflow, Supabase skill refresh, and the setup:verify 401 (URL/anon key mismatch) diagnostic. `AGENTS.md` gained a Skill Routing entry for `docs/ai/skills/react-doctor/SKILL.md.`

**Comment by musab03** [comment]
Merge conflict resolved by keeping the PR's skills/setup narrative and folding in `epic's` factual updates (`scripts/verify/index.mjs`, `dev:donor` single-app hint, combined Key Conventions wording).

File: scripts/sync-agent-skills.mjs [hunk 1]
```diff
@@ -27,6 +37,72 @@ async function overlayDirectory(sourceDir, targetDir) {
...
+}
+
+async function pruneStaleCanonicalSkills(targetRoot, canonicalSkills) {
+  const previous = await readCanonicalManifest(targetRoot);
+  const currentSkillSet = new Set(canonicalSkills);
+  const staleSkills = previous.canonicalSkills.filter(
+    (skillName) => !currentSkillSet.has(skillName),
+  );
+
+  for (const skillName of staleSkills) {
```
[the remark this answers is not in the document]
Comment (musab03) [reply scripts/sync-agent-skills.mjs:139]: Agreed. The manifest is committed JSON and can be merge-corrupted or hand-edited, so treating it as trusted input is not safe. This branch now validates all canonical skill names against a single-segment slug allowlist and asserts that path.resolve(mirrorRoot, name) stays under the mirror root before rm is called. A Vitest test for a ".." manifest entry is included in tests/unit/script-verifiers.test.ts. No separate cherry-pick branch is needed.

File: scripts/sync-agent-skills.mjs [hunk 2]
```diff
@@ -27,6 +37,72 @@ async function overlayDirectory(sourceDir, targetDir) {
...
+    const targetDir = path.join(targetRoot, skillName);
+    await rm(targetDir, { recursive: true, force: true });
```
[the remark this answers is not in the document]
Comment (musab03) [reply scripts/sync-agent-skills.mjs:143]: Addressed in commit `f615142`. The script now validates skill names against `^[a-z0-9]+(?:-[a-z0-9]+)*$` (single-segment slug) and checks resolved paths stay under the mirror root before any rm operation. A unit test covering `".."` in the manifest was added to `tests/unit/script-verifiers.test.ts.
`

File: scripts/sync-agent-skills.mjs [hunk 3]
```diff
@@ -27,6 +37,72 @@ async function overlayDirectory(sourceDir, targetDir) {
...
+
+    return {
+      version:
+        typeof parsed?.version === "number"
+          ? parsed.version
+          : CANONICAL_MANIFEST_VERSION,
+      canonicalSkills: canonicalSkills.sort(),
+    };
+  } catch (error) {
+    const errorCode =
```
[the remark this answers is not in the document]
Comment (musab03) [reply scripts/sync-agent-skills.mjs:94]: Fair point, the duplication is real. We'll track a small follow-up to extract a getNodeErrorCode() helper; it's maintainability only and not tied to this PR's safety goals, so deferring keeps this PR focused. The PR is otherwise ready.

**Comment by musab03** [comment]
> ### 1. Summary
> This PR is almost entirely **developer tooling and documentation** (skill mirror sync, setup scripts, README). It does **not** change Next.js, Supabase, or production app runtime code, so classic production edge cases (auth, RLS, cache tags, etc.) largely **do not apply**.
> 
> Overall the direction is sound: a committed manifest plus `skills:verify` in setup catches **mirror drift** early, which improves **team trust and CI consistency**.
> 
> There is **one high-confidence safety issue** in the new prune logic: **manifest entries are not validated as plain skill directory names**, so a malformed or malicious manifest could make `rm` target **paths outside** the intended skill mirror folder. Everything else is lower severity (documentation drift, optional hardening).
> 
> **Verdict:** Edge cases found; **one change is worth making before merge** (path-safe validation for manifest-driven deletes).
> 
> ### 2. Highest-risk edge cases
> #### Title: Manifest-driven prune can escape the skill mirror directory (`..` / path segments)
> * **Severity:** High
> * **File and line range:** `scripts/sync-agent-skills.mjs` — `readCanonicalManifest` returns unvalidated strings from `parsed.canonicalSkills` (approximately **lines 44–60**), and `pruneStaleCanonicalSkills` uses them in `path.join(targetRoot, skillName)` then `rm` (approximately **lines 92–103**).
> * **Trigger scenario:** A bad merge, hand-edit, or malicious commit puts a manifest entry such as `".."` or `"foo/../../../somewhere"` into `.repo-canonical-skills.json` under `.agents/skills` or `.cursor/skills`. The next `bun run skills:sync` (or anything that runs this script) executes prune.
> * **Why this is realistic:** The manifest is **JSON on disk** in the repo; merge conflicts and accidental edits happen in real teams. The code **trusts** `canonicalSkills` strings for filesystem deletes.
> * **What the code does now:** `staleSkills` is built from the previous manifest; for each stale name it `path.join(targetRoot, skillName)` and `rm(..., { recursive: true, force: true })`. Node `path.join` **normalizes** `..`, so `".."` under `.../.agents/skills` resolves to **`.../.agents`**, not a subdirectory named `..`.
> * **Failure mode:** **Destructive delete outside the intended subtree** (worst case: removing a large portion of `.agents/` or `.cursor/`, not just one skill). This is worse than “stale mirror files” because it is **silent data loss** in the working tree until noticed via git.
> * **Minimal recommended fix:** Before pruning (and optionally before writing the manifest), **reject** any skill name that is not a single path segment safe ID, e.g. match `^[a-z0-9][a-z0-9-]*$` / `^[A-Za-z0-9][A-Za-z0-9_-]*$/` aligned with actual skill folder naming, and **reject** `path.sep`, `..`, absolute paths, empty strings, `.`, `..`. Fail fast with a clear error instead of calling `rm`.
> * **Suggested targeted test:** Extend `tests/unit/script-verifiers.test.ts` with a temp repo that writes `.agents/skills/.repo-canonical-skills.json` containing `"canonicalSkills": [".."]` (or a harmless traversal that resolves inside the temp repo only) and assert the script **exits non-zero** without deleting outside `skills/`. Keep the temp root disposable and assertions strict.
> * **Technical explanation:** `path.join` normalization means **string skill names are not just “folder names”**; they are **relative path operands**. `rm` with `recursive: true` follows the resolved path. Without validation, the manifest becomes a **capability to delete arbitrary relative locations** from the mirror roots.
> * **Plain language explanation:** The new “prune old skills” feature reads a small JSON file and deletes folders based on the names inside it. If that file ever contains something like “go up one folder” instead of a normal skill name, the cleanup step could delete **far more than one skill folder**—it could delete big chunks of the project’s hidden config folders. The fix is to only allow **simple, safe names** (like real skill folder names) and stop with an error if anything else appears.
> 
> ### 3. Additional brittle spots
> 1. **Comment vs. behavior drift (Medium / documentation):** `syncCanonicalSkill` still says “Non-destructive sync” (approximately **lines 153–155**), but the PR adds **destructive pruning** for paths previously recorded in the manifest. That can surprise contributors who manually dropped files under `.agents/skills/<name>` for a skill **removed** from `docs/ai/skills/`. Not a prod bug; worth aligning the comment to “overlay + manifest-tracked prune for removed canonical skills”.
> 2. **Invalid JSON in manifest (Low / fail-closed is OK):** Broken manifest causes `readCanonicalManifest` to throw (non-ENOENT). Sync fails hard with a wrapped error—good. Optional: mention “fix or delete `.repo-canonical-skills.json`” in the error text for faster recovery after conflict markers.
> 3. **Setup now fails on skill drift before `setup:verify` (Low / intentional):** `scripts/setup/index.sh` **lines 128–131** and `scripts/setup.ps1` **lines 185–190** exit non-zero when `skills:verify` fails. That is desirable for consistency, but the **first-run** story is: users may hit drift immediately after pull until they run `skills:sync` and commit—README already calls this out (**README.md** around **lines 8–15**), which helps.
> 
> ### 4. Recommended changes before merge
> 1. **Add strict validation** for manifest-driven and manifest-written skill names in `scripts/sync-agent-skills.mjs` (see Highest-risk item). This is the only change I would treat as **merge-blocking** from an edge-case/safety perspective.
> 2. **Optional:** Adjust the “Non-destructive sync” comment to match pruning semantics.
> 3. **Optional:** Add the small Vitest regression for traversal / invalid names.
> 
> ### 5. Watchlist
> * **Concurrent runs of `skills:sync`:** Two processes in one worktree could interleave writes; unlikely in normal dev. If it becomes an issue, consider documenting “single writer” or adding a cheap lock file—**low confidence** that this matters in practice.
> * **Symlink / junction layouts:** Existing `realpath` guards for agent→cursor mirroring remain; pruning uses `join`+`rm` on the literal skill directory name—**assumption:** mirrors are normal directories; exotic symlink setups could behave unexpectedly. **Low confidence** for this repo’s documented workflows.
> 
> ### 6. Final verdict
> **Changes recommended before merge for edge-case safety** — specifically **validate manifest skill names before any `rm`**.
> 
> ### Optional Cursor follow-up prompt
> ```
> In scripts/sync-agent-skills.mjs (readCanonicalManifest ~44-60, pruneStaleCanonicalSkills ~92-103, writeCanonicalManifest ~81-89, main ~232-245), harden manifest handling: reject canonical skill names that are not safe single path segments (forbid "..", separators, "."-only names, empty strings). Fail sync with a clear error instead of rm. Preserve intended behavior for normal skill folders from docs/ai/skills. Add a focused Vitest in tests/unit/script-verifiers.test.ts that writes a temp .repo-canonical-skills.json containing ".." and asserts the script exits non-zero without deleting outside the temp skills directory. Update the "Non-destructive sync" comment in syncCanonicalSkill (~153-155) only if needed to match pruning semantics. No unrelated refactors.
> ```
> 
> [Open in Web](https://cursor.com/agents/bc-931d46df-256c-4b4f-bfbb-795b2d430ca0) [View Automation](https://cursor.com/automations/6ff02c62-0da9-4eda-88d5-970e688b3614) 
> Sent by Cursor Automation: Edge Cases Test

Fixed in commit `f615142`. Validation now rejects any skill name containing .., path separators, or anything other than lowercase alphanumeric + hyphens. The resolved path containment check adds defense-in-depth. Test added in `tests/unit/script-verifiers.test.ts.`
